src/binary/node.rs

```diff
@@ -1,36 +1,65 @@
-use crate::binary::{leaf_sum, node_sum};
-use crate::common::{Bytes32, Position};
+use crate::{
+    binary::{
+        buffer::{Buffer, ReadView, WriteView, DEFAULT_BUFFER},
+        leaf_sum, node_sum,
+    },
+    common::{Bytes32, Position},
+};
 
 use core::fmt::Debug;
 
 #[derive(Clone, PartialEq, Eq, Debug)]
 pub struct Node {
-    position: Position,
-    hash: Bytes32,
+    buffer: Buffer,
 }
 
 impl Node {
     pub fn create_leaf(index: u64, data: &[u8]) -> Self {
-        let position = Position::from_leaf_index(index);
-        let hash = leaf_sum(data);
-        Self { position, hash }
+        let mut buffer = *DEFAULT_BUFFER;
+        let mut view = WriteView::new(&mut buffer);
+        *view.position_mut() = Position::from_leaf_index(index);
+        *view.hash_mut() = leaf_sum(data);
+        Self { buffer }
     }
 
     pub fn create_node(left_child: &Self, right_child: &Self) -> Self {
-        let position = left_child.position().parent();
-        let hash = node_sum(left_child.hash(), right_child.hash());
-        Self { position, hash }
+        let mut buffer = *DEFAULT_BUFFER;
+        let mut view = WriteView::new(&mut buffer);
+        *view.position_mut() = left_child.position().parent();
+        *view.hash_mut() = node_sum(left_child.hash(), right_child.hash());
+        Self { buffer }
     }
 
     pub fn position(&self) -> Position {
-        self.position
+        let view = ReadView::new(&self.buffer);
+        view.position()
     }
 
     pub fn key(&self) -> u64 {
         self.position().in_order_index()
     }
 
     pub fn hash(&self) -> &Bytes32 {
-        &self.hash
+        let view = ReadView::new(&self.buffer);
+        let ptr = view.hash() as *const Bytes32;
```

Comment:
Why not simply return `view.hash()`?

The reference is owned by the current function, even though the referenced data is owned by the instance and has a sufficient lifetime; unfortunately, there is no way (that I know of) to communicate that to the compiler here without `unsafe`.

Alternatively, we could just return an owned copy of the `Bytes32` and avoid the unsafe code, but I think that returning a reference results in a better, more intuitive interface for the `Node`.